collector/generatorreceiver/internal/flags/flag.go

```diff
@@ -15,109 +12,108 @@ const (
 )
 
 // TODO: separate config types from code types generally
-// TODO: make Flag an interface
+
+type IncidentConfig struct {
+	ParentFlag string        `json:"parentFlag" yaml:"parentFlag"`
+	Start      time.Duration `json:"start" yaml:"start"`
+	End        time.Duration `json:"end" yaml:"end"`
+}
+
+type CronConfig struct {
+	Start string `json:"start" yaml:"start"`
+	End   string `json:"end" yaml:"end"`
+}
 
 type Flag struct {
-	Name          string `json:"name" yaml:"name"`
-	IncidentName  string `json:"incident" yaml:"incident"`
-	Start         string `json:"start" yaml:"start"`
-	End           string `json:"end" yaml:"end"`
-	state         float64
-	cron          *cron.Cron
-	incident      *Incident
-	incidentStart time.Duration
-	incidentEnd   time.Duration
+	Name     string          `json:"name" yaml:"name"`
+	Incident *IncidentConfig `json:"incident" yaml:"incident"`
+	Cron     *CronConfig     `json:"cron" yaml:"cron"`
+
+	started time.Time
+}
+
+func (f *Flag) Active() bool {
+	f.update()
+	return f.active()
 }
 
-func (f *Flag) Enabled() bool {
-	return f.GetState() > DisabledState
+func (f *Flag) active() bool {
+	return !f.started.IsZero()
 }
 
-func (f *Flag) GetState() float64 {
-	if f.incident == nil {
-		return f.state
+// update checks if the given flag f has a parent flag ("Incident"); if so,
+// updates f's state based on its start and end times relative to the parent.
+func (f *Flag) update() {
+	if f.Incident == nil {
+		// managed by cron or manual-only
+		return
+	}
+	// Managed by parent
+	parent := Manager.GetFlag(f.Incident.ParentFlag)
+	if parent == nil {
+		// TODO: with validation, this should never happen
+		return
 	}
 
-	// TODO: all this logic probably doesn't belong here
-	if f.Name == fmt.Sprintf("%s.default", f.incident.name) {
-		if !f.incident.Active() {
-			return EnabledState
-		}
-		return DisabledState
+	incidentDuration := parent.CurrentDuration()
+	afterStart := incidentDuration > f.Incident.Start
+	beforeEnd := f.Incident.End == 0 || incidentDuration < f.Incident.End
+	// shouldBeActive will be true if and only if both of the following are true:
+	// - parent has been Active for at least f's incident start time
+	// - either f has no end *or* parent has been active for less than f's incident end time
+	shouldBeActive := afterStart && beforeEnd
+
+	if f.active() != shouldBeActive {
+		f.Toggle()
```

Comment:
Should this be guarded with a mutex or are flags guaranteed to be thread-safe?

Reply:
Ah, yeah - as I was wrapping this up I realized flags probably do need some synchronization and made a mental todo about it. It seems unlikely to cause serious problems but it might cause confusion. I'll add that.